qiskit_dynamics/perturbation/perturbation_results.py

```diff
@@ -0,0 +1,76 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+r"""
+Class for storing results of perturbation theory computations.
+"""
+
+from typing import List, Union
+from dataclasses import dataclass
+
+from multiset import Multiset
+
+from qiskit import QiskitError
+from qiskit_dynamics.array import Array
+
+
+@dataclass
```

Comment:
Maybe this should be a `frozen=True` dataclass so user cant change the fields once its returned?

Reply:
Yeah good idea - changing it to this.

Reply:
Okay I've ended up changing this, which I think simplifies it a bit. I've made a base class called `_LabeledData` for indexing a list of data with arbitrary labels. Obviously this is dictionary-like but it's nice to not have to make labels hashable, and the class can hold metadata.

There are then two subclasses `PowerSeriesData` and `DysonLikeData` - all they do functionally is make explicit assumptions about what the labels are (`Multiset` or `list`), and what the stored data is (it should be returned as an `Array`).

I still haven't set these to frozen or anything.

I think these changes make the functioning of these data classes a bit clearer. Either way it shouldn't be too important.